pkg/apis/metrics/types.go

```diff
@@ -16,14 +16,205 @@ limitations under the License.
+	// The "core" unit can be interpreted as CPU core-seconds per second.
+	TotalCores *resource.Quantity `json:"totalCores,omitempty"`
+	// CPU load that the container is experiencing, represented as a smoothed average of number of
+	// runnable threads.  Load is averaged over the sampling window.
+	LoadAverage *resource.Quantity `json:"loadAverage,omitempty"`
+}
+
+// MemoryMetrics contains data about memory usage.
+type MemoryMetrics struct {
+	// Total memory in use. This includes all memory regardless of when it was accessed.
+	TotalBytes *resource.Quantity `json:"totalBytes,omitempty"`
+	// The amount of working set memory. This includes recently accessed memory,
+	// dirty memory, and kernel memory. UsageBytes is <= TotalBytes.
+	UsageBytes *resource.Quantity `json:"usageBytes,omitempty"`
+	// Cumulative number of minor page faults.
+	PageFaults *int64 `json:"pageFaults,omitempty"`
+	// Cumulative number of major page faults.
+	MajorPageFaults *int64 `json:"majorPageFaults,omitempty"`
+}
+
+// CustomMetricType specifies the type of metric being exported.
+type CustomMetricType string
+
+const (
+	// Instantaneous value. May increase or decrease.
+	CustomMetricGauge CustomMetricType = "Gauge"
+	// A counter-like value that increases monotonically.
+	CustomMetricCumulative CustomMetricType = "Cumulative"
+	// Rate over a time period.
+	CustomMetricDelta CustomMetricType = "Delta"
+)
+
+// CustomMetric provides custom application metric data.
+// See https://github.com/google/cadvisor/blob/master/docs/application_metrics.md
+type CustomMetric struct {
+	// The name of the metric.
+	Name string `json:"name"`
+	// Type of the metric.
+	Type CustomMetricType `json:"type"`
+	// Display Unit for the sample values.
+	Unit string `json:"unit"`
+	// Historical custom metric samples.
+	Samples []CustomMetricSample `json:"samples,omitempty" patchStrategy:"merge" patchMergeKey:"sampleTime"`
+}
+
+// CustomMetric contains a single sample point for a custom metric.
+type CustomMetricSample struct {
+	Sample `json:",inline"`
+	// Label associated with a metric
+	Label *string `json:"label,omitempty"`
+	// The value of the metric at this point.
+	// The unit is defined by the CustomMetric this sample belongs to.
+	Value resource.Quantity `json:"value"`
+}
+
+// RawMetricsOptions are the query options for raw metrics endpoints.
+type RawMetricsOptions struct {
+	// Only include samples with sampleTime equal to or more recent than this time.
+	// This does not affect cumulative values, which are cumulative from object creation.
+	SinceTime *unversioned.Time `json:"sinceTime,omitempty"`
+	// Only include samples with sampleTime equal to or less recent than this time.
+	UntilTime *unversioned.Time `json:"untilTime,omitempty"`
+	// Specifies the maximum number of elements in any list of samples.
+	// When the total number of samples exceeds the maximum the most recent samples are returned.
+	MaxSamples int `json:"maxSamples,omitempty"`
```

Comment:
For maxSample > 1 there should be a step so that 1 can specify that 1 want 3 samples from the last minute, spaced by 20 sec. Returning 3 most recent samples for 1 min window is not particularly useful.